enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,237 @@
+   (org.apache.logging.log4j Level)
+   (org.apache.logging.log4j.core.appender AbstractAppender FileAppender)
+   (org.apache.logging.log4j.core.config AbstractConfiguration LoggerConfig)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *in-tests* "Set when executed in tests to prevent async removal of files" false)
+
+;;; Request callbacks
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [data callback]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (if *in-tests*
+        (callback)
+        (future (callback))))
+
+    ;; mt/user-http-request goes here
+    clojure.java.io.IOFactory
+    (make-input-stream [_ _]
+      (let [res (io/input-stream (if (instance? File data)
+                                   (ba-copy data)
+                                   data))]
+        (callback)
+        res))))
+
+;;; Storage
+
+(def parent-dir "Dir for storing serialization API export-in-progress and archives."
+  (let [f (io/file (System/getProperty "java.io.tmpdir") "serdesv2-api")]
+    (.mkdirs f)
+    (.deleteOnExit f)
+    (.getPath f)))
+
+(defn- compress-tgz [^File src ^File dst]
+  (when-not (.exists src)
+    (throw (ex-info (format "Path is not readable or does not exist: %s" src)
+                    {:path src})))
+  (let [prefix (.getPath (.getParentFile src))]
+    (with-open [tar (-> (io/output-stream dst)
+                        (GzipCompressorOutputStream. (doto (GzipParameters.)
+                                                       (.setModificationTime (System/currentTimeMillis))))
+                        (TarArchiveOutputStream. 512 "UTF-8"))]
+      (.setLongFileMode tar TarArchiveOutputStream/LONGFILE_POSIX)
+
+      (doseq [^File f (file-seq src)
+              :let    [path-in-tar (subs (.getPath f) (count prefix))
+                       entry (TarArchiveEntry. f path-in-tar)]]
+        (.putArchiveEntry tar entry)
+        (when (.isFile f)
+          (with-open [s (io/input-stream f)]
+            (io/copy s tar)))
+        (.closeArchiveEntry tar)))
+    dst))
+
+(defn- entries [^TarArchiveInputStream tar]
+  (lazy-seq
+   (when-let [entry (.getNextEntry tar)]
+     (cons entry (entries tar)))))
+
+(defn- uncompress-tgz [^File archive ^File dst]
+  (let [dir-name (atom nil)]
+    (with-open [tar (-> (io/input-stream archive)
+                        (GzipCompressorInputStream.)
+                        (TarArchiveInputStream.))]
+      (doseq [^TarArchiveEntry e (entries tar)]
+        (when-not @dir-name
+          (reset! dir-name (.getName e)))
+        (let [f (io/file dst (.getName e))]
+          (if (.isFile e)
+            (io/copy tar f)
+            (.mkdirs f)))))
+    @dir-name))
```

Comment:
sorry but I'm a bit nitty with the use of atom.

this is a version that doesn't use atom, it also a bit faster because we dont' need to deref `dir-name` in the loop.


```clojure
(defn- uncompress-tgz [^File archive ^File dst]
  (with-open [tar (-> (io/input-stream archive)
                      (GzipCompressorInputStream.)
                      (TarArchiveInputStream.))]
    (let [tar-entries (entries tar)]
      (doseq [^TarArchiveEntry e tar-entries]
        (let [f (io/file dst (.getName e))]
          (if (.isFile e)
            (io/copy tar f)
            (.mkdirs f))))
      (when-some [f (first tar-entries)]
        (.getName f)))))
```

Reply:
I don't think this will work since tar as a stream is consumed already, so you'll have to seek back/restart this, and derefing will be much faster than that. Even so, with i/o taking place here performance of derefing is going to be negligible and it's much simpler than seeking back consumed streams.

Reply:
if I understand correctly, `tar-entries` is a lazy-seq and it'll be cached the second time you read from it. so it should work. 

in fact I tried it and it works for me.